reactive/kubernetes_master.py

```diff
@@ -180,11 +180,27 @@ def maybe_install_kube_proxy():
         calculate_and_store_resource_checksums(checksum_prefix, snap_resources)
 
 
+@when_not('kubernetes-master.first-init.ran')
+def first_init():
+    # Do anything that we want to only do once.
+    if not is_state('kubernetes-master.cluster-name.legacy'):
+        # For new installs, this will run before check_for_upgrade_needed,
+        # but for upgrades from older code, that will run first.
+        set_state('kubernetes-master.cluster-name.new')
+    set_state('kubernetes-master.first-init.ran')
+
+
 @hook('upgrade-charm')
 def check_for_upgrade_needed():
     '''An upgrade charm event was triggered by Juju, react to that here.'''
     hookenv.status_set('maintenance', 'Checking resources')
 
+    # Handle upgrade path for cluster name being sent to cdk-addons.
+    if not is_state('kubernetes-master.cluster-name.new'):
+        # For new installs, first_init will run before this, but for
+        # upgrades from older code, this will run first.
```

Comment:
You sure `check_for_upgrade_needed` will run before `first_init`?  Does @hook always happen before reactive handlers?

Reply:
Yes, `@hook` always runs before any `@when` handlers.  Still, I'm not super happy with this, as it feels too implicit / fragile.  But I couldn't think of another way to do it.